Better print

mdinger opened the pull request:
Aims at fixing #495 by adding more detail about printing. Don't merge.

I gotta go do other things now but this is what I've got so far. May not get back to it for a few days.

First commit was a driveby fix.

rust-highfive commented:
r? @steveklabnik

(rust_highfive has picked a reviewer for you, use r? to override)

mdinger commented:
cc @crumblingstatue You may be interested in this.

GBGamer commented:
Wow, this is awesome. :+1:

crumblingstatue commented on examples/print/print.rs:
```diff
@@ -1,23 +1,34 @@
 
-    // Error! You are missing an argument.
-    println!("My name is {0:?}, {1:?} {0:?}", "Bond");
-    // FIXME ^ Add the missing argument: "James"
```
I don't think you should remove the FIXME here. It's helpful to tell the user exactly what he needs to do to fix this problem.

steveklabnik replied:
agreed

mdinger replied:
Oh, I botched the build somewhere and I think I deleted this while tracking down the error. Forgot to add it back.

crumblingstatue commented on examples/print/print.rs:
```diff
@@ -1,23 +1,34 @@
+// Create a structure which contains an `i32`. The structures
+// name is `FailedPrinter`.
+struct FailedPrinter(i32);
```
This feels confusing and out of place to me here at the top.
It would be better to declare this at the point we are trying to print it.
I also think `FailedPrinter` is a confusing name for it. It's not a printer. It's a custom type we're trying to print.

How about something like this instead?
```
// However, custom types such as this structure containing an
// `i32` require more complicated handling. This will
// not work.
struct MyCustomStruct(i32);
println!("This struct `{}` won't print...", MyCustomStruct(3));
// ^ FIXME: Comment this line out
```

mdinger replied:
I think your fix is sensible.

mdinger commented:
Pushed updates. Was working on this offline so I'm not sure how it gels with the previous comments. I'll have to compare again later.

I tried adding `print!` and `format!` but they're so similar that it didn't seem worth the extra space. In general, long examples aren't good so I dropped them. I've still gotta look it over again though.

steveklabnik commented:
Cool, when would you like me to review this?

mdinger commented:
I wanna sleep on it and look at it again tomorrow. I'll try to ping you tomorrow or the next day for review. I think it'll be ready then. I don't think there will be many big changes though. I'll update title from WIP too (gotta remember to do it...)

mdinger commented:
Something came up. Haven't gotten to it yet. I'll try to get to it in a day or so.

mdinger commented:
@steveklabnik r?
I rebased onto that other pull so I could build it. I don't know what it'll do if you push the other first. I assume it'll work though. I don't have any particularly strong criticism remaining about it other than `Display` is kinda long and I don't see much sense in splitting it into more sections.

@crumblingstatue I think I moved that specific `struct` you commented about and I tried to improve the naming in general though they follow naming schemes that makes sense to me. In general, I try to keep `struct` definitions outside of `main()` and rustbyexample historically does that too. I don't have any issues moving them inside where it makes sense though.

As before, comments welcome.

steveklabnik commented:
> I rebased onto that other pull so I could build it. I don't know what it'll do if you push the other first. I assume it'll work though

Oh, I didnt see this till now...

mdinger commented:
You want me to fix it? I'm kinda curious what it'll do...

steveklabnik commented:
It looks like everything is fine, right?

mdinger commented:
I think so. Since the first 3 patches already merged, it should just ignore them and add only the last. That's speculation though. If you don't want to risk it, I can rebase again so there's only a single patch.

Also, rather than closing #495 if you merge, can you check all but the last box. Or leave it to me. I'll do it.

mdinger commented:
@steveklabnik I went ahead and rebased to tip because it makes the PR cleaner though the commit/merge history shouldn't be different either way.

steveklabnik commented:
Okay! Let's land this. Thanks.

mdinger commented:
Cool
